lib/Modules/ADCharmUtils/ADCharmUtils.psm1

```diff
@@ -79,97 +91,141 @@ function Get-MyADCredentials {
-            } else {
-                $ctx["adcredentials"] = $null
-            }
+    # replace the djoin data key with something less dynamic
+    $djoinData = $ctx[$blobKey]
+    $ctx.Remove($blobKey)
+    [string]$ctx["djoin_blob"] = $djoinData
+
+    # Deserialize credential info
+    if($ctx["adcredentials"]) {
+        $creds = Get-MyADCredentials -Credentials $ctx["adcredentials"] -Domain $ctx["netbiosname"]
+        if($creds) {
+            [array]$ctx["adcredentials"] = $creds
+        } else {
+            $ctx["adcredentials"] = $null
         }
-        return $ctx
     }
+    return $ctx
 }
 
-function Invoke-Djoin {
-    [CmdletBinding()]
+function Invoke-DJoin {
     Param(
         [Parameter(Mandatory=$true)]
-        [hashtable]$params
+        [HashTable]$Params
     )
-    PROCESS {
-        Write-JujuInfo "Started Join Domain"
 
-        $networkName = (Get-MainNetadapter)
-        Set-DnsClientServerAddress -InterfaceAlias $networkName -ServerAddresses $params["address"]
-        $cmd = @("ipconfig", "/flushdns")
-        Invoke-JujuCommand -Command $cmd
+    Write-JujuInfo "Started join domain"
 
-        if($params["djoin_blob"]){
-            $blobFile = Join-Path $env:TMP "djoin-blob.txt"
-            Write-FileFromBase64 -File $blobFile -Content $params["djoin_blob"]
-            $cmd = @("djoin.exe", "/requestODJ", "/loadfile", $blobFile, "/windowspath", $env:SystemRoot, "/localos")
-            Invoke-JujuCommand -Command $cmd
-            Invoke-JujuReboot -Now
-        }
+    $networkName = (Get-MainNetadapter)
+    Set-DnsClientServerAddress -InterfaceAlias $networkName -ServerAddresses $Params["address"]
+    $cmd = @("ipconfig", "/flushdns")
+    Invoke-JujuCommand -Command $cmd
+
+    if($Params["djoin_blob"]) {
+        $blobFile = Join-Path $env:TMP "djoin-blob.txt"
+        Write-FileFromBase64 -File $blobFile -Content $Params["djoin_blob"]
+        $cmd = @("djoin.exe", "/requestODJ", "/loadfile", $blobFile, "/windowspath", $env:SystemRoot, "/localos")
+        Invoke-JujuCommand -Command $cmd
+        Invoke-JujuReboot -Now
     }
 }
 
 function Start-JoinDomain {
     $params = Get-ActiveDirectoryContext
-    if ($params.Count){
+    if ($params.Count) {
         if (!(Confirm-IsInDomain $params['domainName'])) {
-            if (!$params["djoin_blob"] -and $params["already-joined-$computername"]) {
+            if (!$params["djoin_blob"] -and $params["already-joined-$COMPUTERNAME"]) {
                 Throw "The domain controller reports that a computer with the same hostname as this unit is already added to the domain, and we did not get any domain join information."
             }
-            Invoke-Djoin -params $params
-        } else {
-            return $true
+            Invoke-DJoin -Params $params
         }
+        return $true
     }
-    Write-JujuInfo "ad-join returned EMPTY"
+    Write-JujuWarning "ad-join returned EMPTY"
     return $false
 }
+
+function Rename-JujuUnit {
```

Comment:
It is only needed when using OpenStack provider and it only does the renaming when `change-hostname` config option is set to true (default is false).

So, if a deployer uses OpenStack provider he should set `change-hostname` for all the charms joining and AD domain, otherwise multiple computers with the same hostname will try to join a domain.